boto/auth.py

```diff
@@ -411,15 +411,19 @@ def canonical_headers(self, headers_to_sign):
         """
         canonical = []
 
+        normalized_headers = {}
         for header in headers_to_sign:
             c_name = header.lower().strip()
             raw_value = str(headers_to_sign[header])
             if '"' in raw_value:
                 c_value = raw_value.strip()
             else:
                 c_value = ' '.join(raw_value.strip().split())
-            canonical.append('%s:%s' % (c_name, c_value))
-        return '\n'.join(sorted(canonical))
+            normalized_headers[c_name] = c_value
+        
+        for key in sorted(normalized_headers.keys()):
```

Comment:
`.keys()` isn't necessary
Also why not `key, value in normalized_headers.items()`?

Reply:
I feel uncomfortable depending on the dict keys to be sorted, since whether this is the case varies by python version. Iterating over sorted tuples could get us into a similar mess, so I don't think we should use .items(). Updating it to remove keys though.

Reply:
Sorry, I meant `sorted(normalized_headers.items())`